sled-agent/src/services.rs

```diff
@@ -1015,57 +1092,146 @@ impl ServiceManager {
             services,
         };
 
-        self.ensure_switch_zone(Some(request)).await
+        self.ensure_zone(ZoneType::Switch, Some(request)).await
     }
 
     /// Ensures that no switch zone is active.
     pub async fn deactivate_switch(&self) -> Result<(), Error> {
-        self.ensure_switch_zone(None).await
+        self.ensure_zone(ZoneType::Switch, None).await
     }
 
-    // Forcefully initialize a switch zone.
+    /// Ensures that an NTP zone exists with the provided configuration.
+    pub async fn activate_ntp(
+        &self,
+        ntp_zone_ip: Option<Ipv6Addr>,
+        servers: Vec<String>,
+        boundary: bool,
+    ) -> Result<(), Error> {
+        info!(
+            self.inner.log,
+            "Ensure NTP service boundary={} IP={:?} servers={:?}",
+            boundary,
+            ntp_zone_ip,
+            servers
+        );
+
+        let services = vec![ServiceType::NTP { servers, boundary }];
+
+        let mut addresses =
+            if let Some(ip) = ntp_zone_ip { vec![ip] } else { vec![] };
+        addresses.push(Ipv6Addr::LOCALHOST);
+
+        // XXXNTP - A boundary NTP zone requires external connectivity
+        // for outbound NTP and DNS traffic.
+
+        let request = ServiceZoneRequest {
+            id: Uuid::new_v4(),
+            zone_type: ZoneType::NTP,
+            addresses,
+            gz_addresses: vec![],
+            services,
+        };
+
+        self.ensure_zone(ZoneType::NTP, Some(request)).await
+    }
+
+    /// Ensures that no NTP zone is active.
+    pub async fn deactivate_ntp(&self) -> Result<(), Error> {
+        self.ensure_zone(ZoneType::NTP, None).await
+    }
+
+    pub fn boundary_ntp_servers(&self) -> Vec<String> {
+        // XXXNTP - How can we find the scrimlet sled subnets?
+        //
+        // If we can get a list of scrimlet peers from maghemite, then
+        // we can just use those. Otherwise, we may have to attempt to
+        // synchronise with the NTP zone on each peer, and use any
+        // that are willing to provide time to us (this will just be
+        // boundary servers).
+        //
+        // This should be done from the local running NTP zone found
+        // via self.innner.ntp_zone - SledLocalZone::Running { request, zone }
+        //
+        // /usr/sbin/chronyd -t10 -L3 -q 'server <sled prefix>::3 iburst'
+        // Exit status 1, no luck, status 0 means we synchronised (the time
+        // will jump) and we can add this IP to the set.
+        //
+        // This strategy is blocked by at least
+        //   https://github.com/oxidecomputer/omicron/issues/2562
```

Comment:
I'll take a look at this issue, but yeah, we should be able to use maghemite for this.